### ISSUE TITLE
Heap-buffer-overflow in sd-dhcp6-client.c

### ISSUE THREAD

**Issue by mrc0mmand** [systemd/systemd#13578]
Discovered by Fuzzit:

```
FUZZER: ==26==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60600000f41c at pc 0x0000005964f8 bp 0x7ffce1a96d50 sp 0x7ffce1a96d48
FUZZER: READ of size 4 at 0x60600000f41c thread T0
FUZZER: SCARINESS: 17 (4-byte-read-heap-buffer-overflow)
FUZZER:     #0 0x5964f7 in client_parse_message /home/travis/build/systemd/systemd/build/../src/libsystemd-network/sd-dhcp6-client.c:1005:39
FUZZER:     #1 0x593d1a in client_receive_reply /home/travis/build/systemd/systemd/build/../src/libsystemd-network/sd-dhcp6-client.c:1058:13
FUZZER:     #2 0x5916ce in client_receive_message /home/travis/build/systemd/systemd/build/../src/libsystemd-network/sd-dhcp6-client.c:1208:21
FUZZER:     #3 0x7f3ee7880d9f in source_dispatch /home/travis/build/systemd/systemd/build/../src/libsystemd/sd-event/sd-event.c:2828:21
FUZZER:     #4 0x7f3ee787febc in sd_event_dispatch /home/travis/build/systemd/systemd/build/../src/libsystemd/sd-event/sd-event.c:3241:21
FUZZER:     #5 0x7f3ee7882348 in sd_event_run /home/travis/build/systemd/systemd/build/../src/libsystemd/sd-event/sd-event.c:3299:21
FUZZER:     #6 0x5804f1 in fuzz_client /home/travis/build/systemd/systemd/build/../src/fuzz/fuzz-dhcp6-client.c:44:9
FUZZER:     #7 0x580167 in LLVMFuzzerTestOneInput /home/travis/build/systemd/systemd/build/../src/fuzz/fuzz-dhcp6-client.c:56:9
FUZZER:     #8 0x44a55a in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:576:15
FUZZER:     #9 0x449c75 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:485:3
FUZZER:     #10 0x44c0ab in fuzzer::Fuzzer::MutateAndTestOne() /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:713:19
FUZZER:     #11 0x44cd85 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<std::__Fuzzer::basic_string<char, std::__Fuzzer::char_traits<char>, std::__Fuzzer::allocator<char> >, fuzzer::fuzzer_allocator<std::__Fuzzer::basic_string<char, std::__Fuzzer::char_traits<char>, std::__Fuzzer::allocator<char> > > > const&) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:844:5
FUZZER:     #12 0x4435d3 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:765:6
FUZZER:     #13 0x46b2b2 in main /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
FUZZER:     #14 0x7f3ee616c2e0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202e0)
FUZZER:     #15 0x41fb08 in _start (/app/fuzzer+0x41fb08)
FUZZER: 
FUZZER: 0x60600000f41e is located 0 bytes to the right of 62-byte region [0x60600000f3e0,0x60600000f41e)
FUZZER: allocated by thread T0 here:
FUZZER:     #0 0x550ea3 in __interceptor_malloc /tmp/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:146:3
FUZZER:     #1 0x59111d in client_receive_message /home/travis/build/systemd/systemd/build/../src/libsystemd-network/sd-dhcp6-client.c:1137:19
FUZZER:     #2 0x7f3ee7880d9f in source_dispatch /home/travis/build/systemd/systemd/build/../src/libsystemd/sd-event/sd-event.c:2828:21
FUZZER:     #3 0x7f3ee787febc in sd_event_dispatch /home/travis/build/systemd/systemd/build/../src/libsystemd/sd-event/sd-event.c:3241:21
FUZZER:     #4 0x7f3ee7882348 in sd_event_run /home/travis/build/systemd/systemd/build/../src/libsystemd/sd-event/sd-event.c:3299:21
FUZZER:     #5 0x5804f1 in fuzz_client /home/travis/build/systemd/systemd/build/../src/fuzz/fuzz-dhcp6-client.c:44:9
FUZZER:     #6 0x580167 in LLVMFuzzerTestOneInput /home/travis/build/systemd/systemd/build/../src/fuzz/fuzz-dhcp6-client.c:56:9
FUZZER:     #7 0x44a55a in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:576:15
FUZZER:     #8 0x449c75 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:485:3
FUZZER:     #9 0x44c0ab in fuzzer::Fuzzer::MutateAndTestOne() /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:713:19
FUZZER:     #10 0x44cd85 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<std::__Fuzzer::basic_string<char, std::__Fuzzer::char_traits<char>, std::__Fuzzer::allocator<char> >, fuzzer::fuzzer_allocator<std::__Fuzzer::basic_string<char, std::__Fuzzer::char_traits<char>, std::__Fuzzer::allocator<char> > > > const&) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:844:5
FUZZER:     #11 0x4435d3 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:765:6
FUZZER:     #12 0x46b2b2 in main /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
FUZZER:     #13 0x7f3ee616c2e0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202e0)
FUZZER: 
FUZZER: SUMMARY: AddressSanitizer: heap-buffer-overflow /home/travis/build/systemd/systemd/build/../src/libsystemd-network/sd-dhcp6-client.c:1005:39 in client_parse_message
FUZZER: Shadow bytes around the buggy address:
FUZZER:   0x0c0c7fff9e30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
FUZZER:   0x0c0c7fff9e40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
FUZZER:   0x0c0c7fff9e50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
FUZZER:   0x0c0c7fff9e60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
FUZZER:   0x0c0c7fff9e70: fa fa fa fa fa fa fa fa fa fa fa fa 00 00 00 00
FUZZER: =>0x0c0c7fff9e80: 00 00 00[06]fa fa fa fa 00 00 00 00 00 00 00 06
FUZZER:   0x0c0c7fff9e90: fa fa fa fa fd fd fd fd fd fd fd fa fa fa fa fa
FUZZER:   0x0c0c7fff9ea0: fd fd fd fd fd fd fd fa fa fa fa fa fd fd fd fd
FUZZER:   0x0c0c7fff9eb0: fd fd fd fa fa fa fa fa fd fd fd fd fd fd fd fa
FUZZER:   0x0c0c7fff9ec0: fa fa fa fa fd fd fd fd fd fd fd fa fa fa fa fa
FUZZER:   0x0c0c7fff9ed0: fd fd fd fd fd fd fd fa fa fa fa fa fd fd fd fd
FUZZER: Shadow byte legend (one shadow byte represents 8 application bytes):
FUZZER:   Addressable:           00
FUZZER:   Partially addressable: 01 02 03 04 05 06 07 
FUZZER:   Heap left redzone:       fa
FUZZER:   Freed heap region:       fd
FUZZER:   Stack left redzone:      f1
FUZZER:   Stack mid redzone:       f2
FUZZER:   Stack right redzone:     f3
FUZZER:   Stack after return:      f5
FUZZER:   Stack use after scope:   f8
FUZZER:   Global redzone:          f9
FUZZER:   Global init order:       f6
FUZZER:   Poisoned by user:        f7
FUZZER:   Container overflow:      fc
FUZZER:   Array cookie:            ac
FUZZER:   Intra object redzone:    bb
FUZZER:   ASan internal:           fe
FUZZER:   Left alloca redzone:     ca
FUZZER:   Right alloca redzone:    cb
FUZZER:   Shadow gap:              cc
FUZZER: ==26==ABORTING

```

Fuzzer input which caused the crash:
[test-dhcp6-client-crash.tar.gz](https://github.com/systemd/systemd/files/3621316/test-dhcp6-client-crash.tar.gz)

cc @yuwata

**Comment by mrc0mmand** [comment]
MSan report from Fuzzit:

```
FUZZER: ==28==WARNING: MemorySanitizer: use-of-uninitialized-value
FUZZER:     #0 0x53eac4 in be32toh /home/travis/build/systemd/systemd/build/../src/basic/sparse-endian.h:86:55
FUZZER:     #1 0x53c561 in client_parse_message /home/travis/build/systemd/systemd/build/../src/libsystemd-network/sd-dhcp6-client.c:1005:31
FUZZER:     #2 0x53b0b3 in client_receive_reply /home/travis/build/systemd/systemd/build/../src/libsystemd-network/sd-dhcp6-client.c:1058:13
FUZZER:     #3 0x5385de in client_receive_message /home/travis/build/systemd/systemd/build/../src/libsystemd-network/sd-dhcp6-client.c:1208:21
FUZZER:     #4 0x7fd4f2611602 in source_dispatch /home/travis/build/systemd/systemd/build/../src/libsystemd/sd-event/sd-event.c:2828:21
FUZZER:     #5 0x7fd4f26104c2 in sd_event_dispatch /home/travis/build/systemd/systemd/build/../src/libsystemd/sd-event/sd-event.c:3241:21
FUZZER:     #6 0x7fd4f2612d30 in sd_event_run /home/travis/build/systemd/systemd/build/../src/libsystemd/sd-event/sd-event.c:3299:21
FUZZER:     #7 0x520af2 in fuzz_client /home/travis/build/systemd/systemd/build/../src/fuzz/fuzz-dhcp6-client.c:44:9
FUZZER:     #8 0x5202ae in LLVMFuzzerTestOneInput /home/travis/build/systemd/systemd/build/../src/fuzz/fuzz-dhcp6-client.c:56:9
FUZZER:     #9 0x44a6da in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:576:15
FUZZER:     #10 0x449df5 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:485:3
FUZZER:     #11 0x44c22b in fuzzer::Fuzzer::MutateAndTestOne() /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:713:19
FUZZER:     #12 0x44cf05 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<std::__Fuzzer::basic_string<char, std::__Fuzzer::char_traits<char>, std::__Fuzzer::allocator<char> >, fuzzer::fuzzer_allocator<std::__Fuzzer::basic_string<char, std::__Fuzzer::char_traits<char>, std::__Fuzzer::allocator<char> > > > const&) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:844:5
FUZZER:     #13 0x443753 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:765:6
FUZZER:     #14 0x46b432 in main /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
FUZZER:     #15 0x7fd4f10062e0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202e0)
FUZZER:     #16 0x41fc88 in _start (/app/fuzzer+0x41fc88)
FUZZER: 
FUZZER:   Uninitialized value was created
FUZZER:     <empty stack>
FUZZER: 
FUZZER: SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/travis/build/systemd/systemd/build/../src/basic/sparse-endian.h:86:55 in be32toh
```

**Comment by yuwata** [comment]
Fix is waiting in #13580.
